### PR TITLE
Fix trace agent connection issues that emerge in Azure App Service environment

### PR DESCRIPTION
**Pull request by zacharycmontoya**
- Fix regression from v1.18.0 when the Tracer began using WebRequest instead of HttpClient. This change caused us to miss SocketExceptions that indicated we needed to restart the trace agent, since the SocketException was nested in two exceptions instead of one. Now, we unwrap the exceptions until reaching the innermost exception to determine if we encountered the SocketException.
- Make Api._tracesEndpoint mutable so retries can receive an updated port number, if needed.
- Reset the ProcessMetadata.HasAttemptedStartup field when forcing a restart so consecutive failed Trace agent responses don't trigger resets that are already in-progress.
- Immediately catch ArgumentExceptions when invoking Process.GetProcessById because we fully expect this to occur.

@DataDog/apm-dotnet

### CODE REVIEW

File: src/Datadog.Trace/Agent/Api.cs [hunk 1]
```diff
@@ -96,27 +101,36 @@ public async Task<bool> SendTracesAsync(Span[][] traces)
...
+                    while (innermostException?.InnerException != null)
                     {
-                        isSocketException = true;
-                        Log.Error(se, "Unable to communicate with the trace agent at {0}", _tracesEndpoint);
-                        TracingProcessManager.TryForceTraceAgentRefresh();
+                        innermostException = innermostException.InnerException;
                     }
 
-                    if (retryCount >= retryLimit)
+                    if (innermostException is SocketException se)
```
Comment (colin-higgins) [inline src/Datadog.Trace/Agent/Api.cs:121]: can we nest this in the loop and break if we detect SocketException? Not that it won't be innermost, but it seems flexible that way.

File: src/Datadog.Trace/TracingProcessManager.cs [hunk 2]
```diff
@@ -229,6 +230,12 @@ private static void InitializePortManagerClaimFiles(string traceAgentDirectory)
                                 }
                             }
                         }
+                        catch (ArgumentException)
```
Comment (colin-higgins) [inline src/Datadog.Trace/TracingProcessManager.cs:233]: Good add

File: src/Datadog.Trace/Agent/Api.cs (hunk 1 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace/Agent/Api.cs]: Good idea, fixed in c4824856a1c5b6323352f525aec7c762d1047247

File: src/Datadog.Trace/Agent/IAgentWriter.cs [hunk 3]
```diff
@@ -12,6 +13,6 @@ internal interface IAgentWriter
 
         Task FlushAndCloseAsync();
 
-        void OverrideApi(IApi api);
+        void OverrideApiBaseEndpoint(Uri uri);
```
Comment (lucaspimentel) [inline src/Datadog.Trace/Agent/IAgentWriter.cs:16]: Consider renaming to `SetApiBaseEndpoint()`.

File: src/Datadog.Trace/Agent/IApi.cs [hunk 4]
```diff
@@ -1,9 +1,12 @@
+using System;
 using System.Threading.Tasks;
 
 namespace Datadog.Trace.Agent
 {
     internal interface IApi
     {
+        public void OverrideBaseEndpoint(Uri baseEndpoint);
```
Comment (lucaspimentel) [inline src/Datadog.Trace/Agent/IApi.cs:8]: Consider renaming to `SetBaseEndpoint()`.

Comment (zacharycmontoya) [reply src/Datadog.Trace/Agent/IApi.cs]: Good idea, implemented in a8ae6f5ec783d182198779a59a280a26eb12a5d1

File: src/Datadog.Trace/Agent/IAgentWriter.cs (hunk 3 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace/Agent/IAgentWriter.cs]: Good idea, implemented in a8ae6f5ec783d182198779a59a280a26eb12a5d1

**Pull request merged**